### ISSUE TITLE
Funnel works only within tailnet

### ISSUE THREAD

**Issue by marcinjahn** [tailscale/tailscale#11849]
### What is the issue?

It seems to me that `funnel` and `serve` do the same thing.
Both allow me to serve content, which is then available via HTTPS only within my tailnet. As I understand, `funnel` should allow me to publish content publicly, on internet. However, my funneled resource works only when I'm connected to my tailnet.

### Steps to reproduce

1. On Device A, Run some web app locally on one of the nodes, e.g. via `podman run -d -p 8080:80 tutum/hello-world`
2. On Device A, Enable funnel: `tailscale funnel 8080`
3. On another Device B (not connected to your tailnet), try to navigate to Device A's DNS address.
4. Page will not be found

### Are there any recent changes that introduced the issue?

Nope, it never worked for me

### OS

Linux, iOS

### OS version

Fedora 39, iOS 17.4.1

### Tailscale version

1.64.0

### Bug report

BUG-e75df7e9a917df61bd3b4c6244e8924e06db465d45115a81747489e4fae37b8c-20240423153805Z-00750e64e97bc4fe

**Comment by jimmybrancaccio** [comment]
I was able to replicate this today due to an incoming support ticket (TSS-37882). I setup a basic `docker-compose.yml`:

```json
version: "3.7"
services:
  ts-dillinger-webserver:
    image: tailscale/tailscale:latest
    container_name: ts-dillinger
    hostname: dillinger
    environment:
      - TS_AUTHKEY=tskey-client-kMq9ax11CNTRL-gGBoK58SWxnGj
      - TS_EXTRA_ARGS=--advertise-tags=tag:container
      - TS_SERVE_CONFIG=/config/dillinger.json
      - TS_STATE_DIR=/var/lib/tailscale
    volumes:
      - /home/jimmy/.docker/data/ts-dillinger-webserver/state:/var/lib/tailscale
      - /home/jimmy/.docker/data/ts-dillinger-webserver/config:/config
      - /dev/net/tun:/dev/net/tun
    cap_add:
      - net_admin
      - sys_module
    restart: unless-stopped
  ts-dillinger:
    image: lscr.io/linuxserver/dillinger:latest
    container_name: dillinger
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=America/Chicago
    volumes:
      - /home/jimmy/.docker/data/dillinger/config:/config
    network_mode: service:ts-dillinger-webserver
    depends_on:
      - ts-dillinger-webserver
    restart: unless-stopped
```

In my SERVE file I have:

```json
{
  "TCP": {
    "443": {
      "HTTPS": true
    }
  },
  "Web": {
    "${TS_CERT_DOMAIN}:443": {
      "Handlers": {
        "/": {
          "Proxy": "http://127.0.0.1:8080"
        }
      }
    }
  },
  "AllowFunnel": {
    "${TS_CERT_DOMAIN}:443": true
  }
}
```

I ran `docker compose up -d --remove-orphans` and everything looks to have started up fine. I observed these couple of lines which seems to indicate the funnel is running:

```
ts-dillinger  | 2024/05/07 20:34:15 serve: closing idle connections to http://127.0.0.1:8080
ts-dillinger  | boot: 2024/05/07 20:34:15 Applying serve config
ts-dillinger  | 2024/05/07 20:34:15 serve: creating a new proxy handler for http://127.0.0.1:8080
```

I see my container show up in my admin console, however I observed in the admin console, clicking on the machine to go to the machine page, under the TLS Configuration section the domain shows properly, but the Status shows as No certificate found. I observed that when trying to visit https://dillinger.taildfbdd.ts.net failed and I also couldn't resolve the hostname.

I then swapped from a different tailnet I was on to the tailnet where I had deployed the container node and I tried bringing up https://dillinger.taildfbdd.ts.net. At this point I observed in the Docker container logs:

```
ts-dillinger  | 2024/05/07 20:43:39 cert("dillinger.taildfbdd.ts.net"): registered ACME account.
ts-dillinger  | 2024/05/07 20:43:39 cert("dillinger.taildfbdd.ts.net"): starting SetDNS call...
ts-dillinger  | 2024/05/07 20:43:50 cert("dillinger.taildfbdd.ts.net"): did SetDNS
ts-dillinger  | 2024/05/07 20:43:51 cert("dillinger.taildfbdd.ts.net"): requesting cert...
ts-dillinger  | 2024/05/07 20:43:52 cert("dillinger.taildfbdd.ts.net"): got cert
ts-dillinger  | 2024/05/07 20:43:52 Accept: TCP{100.65.211.27:65338 > 100.80.14.102:443} 52 tcp non-syn
ts-dillinger  | 2024/05/07 20:43:52 http: TLS handshake error from 100.65.211.27:65338: EOF
dillinger     | 100.65.211.27 - GET / HTTP/1.1 200 - - 329.889 ms
dillinger     | 100.65.211.27 - GET /assets/app-9a52b028cbd397404e2c56df7c2f8e87.css HTTP/1.1 200 72955 - 7.123 ms
dillinger     | 100.65.211.27 - GET /img/icons/file.svg HTTP/1.1 200 - - 12.122 ms
dillinger     | 100.65.211.27 - GET /js/main.bundle.js HTTP/1.1 200 - - 20.965 ms
dillinger     | 100.65.211.27 - GET /theme-github.js HTTP/1.1 404 154 - 27.085 ms
dillinger     | 100.65.211.27 - GET /img/icons/code.svg HTTP/1.1 200 902 - 11.722 ms
dillinger     | 100.65.211.27 - GET /img/icons/enter-zen.svg HTTP/1.1 200 - - 13.855 ms
ts-dillinger  | 2024/05/07 20:44:02 Accept: TCP{100.65.211.27:65339 > 100.80.14.102:443} 107 tcp non-syn
```

so it seems clear that just trying to visit the service triggered a request for a TLS certificate. I had waited about 15 minutes before trying this as [our documentation](https://tailscale.com/kb/1223/funnel#run-tailscale-funnel) indicates it should happen in about 10 minutes but didn't for me.

In summary, I experienced the same issue as described in the original post as well as is the customer who submitted a recent support ticket. Funnel is working more like Serve in that you cannot reach the funneled service if you're not connected to the tailnet where the funnel is running.

**Comment by tommyboylab** [comment]
I'm encountering a similar issue with funnels but haven't yet investigated it as deeply as above. 

Simple issue is I'm hosting a few docker containers on a local machine (also an exit-node) of which I'd like to funnel. Each have an assigned port local to the machine which I can access within the tailnet (both by IP and by the magic-dns). With funnel on I can also access the services from the defined endpoint (either root or from --set-path) but if outside the tailnet this fails.

I've tried creating certs using the tailscale cert <name> command along with restarting the service on the local machine but when visiting the funneled endpoint I get the error from the browser `<server> unexpectedly closed the connection.`  

Without hijacking the issue above if there's a connected problem or missing configuration needed please inform me.

**Comment by brian316** [comment]
i followed this video here from tailscale https://www.youtube.com/watch?v=tqvvZhGrciQ
and used the following docker compose file
```
---
version: "3.7"
services:
  ts-homeassistant:
    image: tailscale/tailscale:latest
    container_name: ts-homeassistant
    hostname: homeassistant
    environment:
      - TS_AUTHKEY=tskey-auth-removed-for-obvious-reasons
      - "TS_EXTRA_ARGS=--advertise-tags=tag:container --reset"
      - TS_STATE_DIR=/var/lib/tailscale
      - TS_SERVE_CONFIG=/config/serve-config.json
      - TS_USERSPACE=false
    volumes:
      - ${PWD}/tailscale-nginx/state:/var/lib/tailscale
      - /dev/net/tun:/dev/net/tun
      - ${PWD}/serveconfig:/config
    cap_add:
      - net_admin
      - sys_module
      - sys_admin
    restart: unless-stopped
    privileged: true
    network_mode: host
  homeassistant:
    image: ghcr.io/home-assistant/home-assistant:latest
    container_name: homeassistant
    environment:
      - TZ=America/Los_Angeles
    volumes:
      - /home/homeassistant:/config
      - /run/dbus:/run/dbus:ro
      - /etc/localtime:/etc/localtime:ro
    restart: unless-stopped
    privileged: true
    depends_on:
      - ts-homeassistant
    network_mode: service:ts-homeassistant

volumes:
  ts-homeassistant:
    driver: local
```

and this is my serve-config.json
```json
{
    "TCP": {
      "443": {
        "HTTPS": true
      }
    },
    "Web": {
      "${TS_CERT_DOMAIN}:443": {
        "Handlers": {
          "/": {
            "Proxy": "http://127.0.0.1:8123"
          }
        }
      }
    },
    "AllowFunnel": {
      "${TS_CERT_DOMAIN}:443": false
    }
  }
```

im not even using the funnel yet and i get an error for acme request.
```
acme: ARI check failed: failed to fetch renewal info from ACME server: fetching renewal info: 400 urn:ietf:params:acme:error:malformed: While parsing ARI CertID an error occurred :: Invalid path; falling back to expiry-based check
```

**Comment by gglanzani** [comment]
For me the issue was that I had `tailnet` also installed on the host. Once removed, things started working.

Weird, though, that everything seemed to be running along happy while fundamentally broken.

**Comment by jimmybrancaccio** [comment]
Can confirm [the Mealie section in our Docker blog article](https://tailscale.com/blog/docker-tailscale-guide#remote-access) isn't working due to the fact that you can't access the hostname which is generated. I've followed the article to a T again today and cannot access the Mealie instance while not being on the tailnet.

From `docker exec -it ts-mealie tailscale status`:

```
# Funnel on:
#     - https://mealie.taildfbdd.ts.net
```

Bringing that URL up in Chrome fails with `ERR_NAME_NOT_RESOLVED`.

**Comment by adriancaruana** [comment]
I also seem to be having the same issue where I can access the container-running funnels on other devices that are connected to my tailnet, but not from the internet. DNS lookups for my MagicDNS name fail, yet the "Machines" tab in my tailscale web admin console, and the command `tailscale funnel status` when run inside my funnel-running container show that the funnel is supposedly working just fine. I'm also getting `ERR_NAME_NOT_RESOLVED` in Chrome when trying to access the funnel from the internet.

**Comment by nkbooth** [comment]
I spent several hours with this annoying me today, and I think I finally solved it for myself.

I made lots of changes, but I think that there are 2 things that fully control outside access via funnel:

In the documentation, there is a config.json file to set up funnel.  The last blob is:
```json
"AllowFunnel": {
    "${TS_CERT_DOMAIN}:443": false
  }
```
The key value should be *true*, not false.

Secondly, ACLs control publishing funnel data to DNS.  Without correct ACL setup, the DNS records will not be updated/set for the funnel, which is what is driving server not found errors - they're correct.  In the ACL, we need to edit the nodeAttrs section to add permission for the tag used on containers to use funnel.  In my case, I'm using the (very original) tag `containers`, so my ACL looks like this:
```json
"nodeAttrs": [
		{"target": ["tag:containers"], "attr": ["funnel"]}
	],
```

Hopefully that helps someone else down the road.

**Comment by adriancaruana** [comment]
Oh, wow, thank you @nkbooth, your second point did it for me. 

Here's be `nodeAttrs` before (not working) and after (working):
```patch
	"nodeAttrs": [
		{
			// Funnel policy, which lets tailnet members control Funnel
			// for their own devices.
			// Learn more at https://tailscale.com/kb/1223/tailscale-funnel/
-                       "target": ["autogroup:member"],
+                       "target": ["autogroup:member", "tag:container"],
			"attr":   ["funnel"],
		},
	],
```

**Comment by jimmybrancaccio** [comment]
Welp, I feel like a dummy! You're totally right @nkbooth! My `nodeAttr` was missing the tag for my containers. Once I added that a DNS record and the certificate were properly fetched and setup. 🤦🏼

**Comment by derrickrc** [comment]
I am currently having the same issue with a Tailscale Funnel I just created on a Linux machine yesterday. I can access the funnel on my Mac which is on the same Tailnet, but only using the http address, e.g: 

`http://<my tailnet url>.ts.net:<port>/endpoint`

But the https full domain just results in a `<url> is currently unable to handle this request. HTTP ERROR 502`

I have an identical Linux machine in my Tailnet located offsite where I am able to access the https funnel, as well as other machines.

**Comment by EDIflyer** [comment]
> ~EDIT: Took 2 days to populate the DNS but seems to be working now, removing comment~

Mine thankfully didn't take that long but it did take about 15 min to become active - your post was very helpful though in recognising that might be the issue, I was getting confused if I'd done something wrong as it was showing on the tailnet OK but not via the public internet.

**Comment by Marerc** [comment]
> Oh, wow, thank you @nkbooth, your second point did it for me.
> 
> Here's my `nodeAttrs` before (not working) and after (working):
> 
> ```diff
> 	"nodeAttrs": [
> 		{
> -                       "target": ["autogroup:member"],
> +                       "target": ["autogroup:member", "tag:container"],
> 			"attr":   ["funnel"],
> 		},
> 	],
> ```

This is a soooo wicked parameter.

**Comment by jeremymeyers** [comment]
> I am currently having the same issue with a Tailscale Funnel I just created on a Linux machine yesterday. I can access the funnel on my Mac which is on the same Tailnet, but only using the http address, e.g:
> 
> `http://<my tailnet url>.ts.net:<port>/endpoint`
> 
> But the https full domain just results in a `<url> is currently unable to handle this request. HTTP ERROR 502`
> 
> I have an identical Linux machine in my Tailnet located offsite where I am able to access the https funnel, as well as other machines.
> 
> **Edit: TS support got back to me, I had specified the wrong port when creating the tunnel. My mistake.**

Could you detail what you did wrong and what fixed it?

**Comment by nachol** [comment]
> Oh, wow, thank you [@nkbooth](https://github.com/nkbooth), your second point did it for me.
> 
> Here's my `nodeAttrs` before (not working) and after (working):
> 
> 	"nodeAttrs": [
> 		{
> -                       "target": ["autogroup:member"],
> +                       "target": ["autogroup:member", "tag:container"],
> 			"attr":   ["funnel"],
> 		},
> 	],

Holy S... spent tons of hours on this one. Thanks a lot mate.

**Comment by derrickrc** [comment]
We are currently having this issue again with a machine in our Tailnet where public requests are failing, but we can still connect to it from our own machines within our Tailnet. The funnel was working fine and accepting public requests up until 2 days ago and it's refused requests since. 

In digging into it a little more, the one thing I see that appears to be different about this machine vs. the others in our Tailnet (whose public funnels are all working fine) is that `IPV6` is `Yes` under client connectivity. All of the other machines have No. 

I filed a ticket with Tailscale, but anyone have any idea what might be going on? Our customer said they rebooted their machine this morning as well and it's still not working.

<img width="554" alt="Image" src="https://github.com/user-attachments/assets/b96c5e79-9bec-4346-995c-1b3260307ba6" />

**Comment by blopker** [comment]
I'm having the same issue as @derrickrc. It seems like the public DNS is not resolving for my ts.net domain. When connected to the Tailnet, `dig` gives me one record. When disconnected I get 0. 

This is for a service that has been working publicly for months, and just stopped working with no changes.

**Comment by imamfahrudin** [comment]
If your Tailscale funnel isn't working as expected, double-check that you've run the command `tailscale funnel enable`. That activates the funnel on your node and makes it available.

**Comment by nikhild64** [comment]
Do we have a solution/workaround to this, i tried today and its not working

**Comment by aroxu** [comment]
Also not working via a funnel. Maybe Tailscale's funnel server is in trouble?

**Comment by KD-MM2** [comment]
same problem today... and nothing of those above works...

**Comment by ananthb** [comment]
Same issue as of yesterday. Both funnelled services went down. Shame there's no status page for this feature.

**Comment by KD-MM2** [comment]
> Same issue as of yesterday. Both funnelled services went down. Shame there's no status page for this feature.

i noticed that my funnel are up again. how about yours?

**Comment by ananthb** [comment]
They were down an hour ago when I had commented, but they are back up now!

**Comment by gerald-lnj** [comment]
same, they are back up

**Comment by pjpscriv** [comment]
I'm having this issue too, for a funnel that has previously worked for months without issue.

I can access the funnel when logged into my Tailnet, but 502 error when I'm logged out.

Hoping that commenting on here will magically resolve my problem as it seems to have for others.

**Comment by akshun-j** [comment]
> I'm having this issue too, for a funnel that has previously worked for months without issue.
> 
> I can access the funnel when logged into my Tailnet, but 502 error when I'm logged out.

Same here  Happened a few hours ago and has yet to resolve.  Console shows connected and funnel is live.

**Comment by akshun-j** [comment]
I guess I should take it all back.  2hrs later, everything is back up again.  That was weird.  I didn't change anything.  Magic, @pjpscriv

**Comment by pjpscriv** [comment]
Lol yep - all working again for me now too. Sometimes the fix really is "just wait a bit" 😅.

**Edit:** I emailed support@tailscale.com about this and and got this reply:

> It looks like this is working now and our team confirms there was some infrastructure updates which ended up inadvertently having impact on availability. Are you seeing them working now too?

So nice to know it was indeed something on their side. Note to self (or anyone else who this happens to), there's a status page that can be checked here: https://status.tailscale.com

**Comment by Odin-byte** [comment]
Is this still an issue for anyone else? I haven't been able to access my funnel outside the Tailscale network for at least two weeks. The funnel and the container are both up and running. The Tailscale status page does not indicate any errors concerning their funnel services.

**Comment by martigr** [comment]
> Is this still an issue for anyone else? I haven't been able to access my funnel outside the Tailscale network for at least two weeks. The funnel and the container are both up and running. The Tailscale status page does not indicate any errors concerning their funnel services.

I observe also issues with funnel for about 2-3 weeks. Most of the time funnel is not working… when connecting to the tailnet, everythinf is fine.

**Comment by electriquo** [comment]
The issue has been open for more than 2 years :(

The [status page](https://status.tailscale.com/) indicates that the Funnel is working. After verifying that all configurations are correct (I can share the details and explain how to confirm that Tailscale honors them), certificates are still not being generated.

Does anyone have any updates?  
Is there anyone from Tailscale monitoring this thread?

**Comment by electriquo** [comment]
Seems that making a HTTPS request to the public hostname (see `tailscale funnel status`) causing to generate the certificates on the fly.

**Comment by arikamidotnet** [comment]
Also experiencing this same issue now as well. worked fine but now all status says operational but only works when connected to tailnet

**Comment by jw-00000** [comment]
For me this stopped working today. The fix was as follows. (Thanks to [nkbookth's comment](https://github.com/tailscale/tailscale/issues/11849#issuecomment-2211623437).)

In the ACL policy I had:

```
"nodeAttrs": [
    {"target": ["100.100.1.100"], "attr": ["funnel"]}
],
```

with 100.100.1.100 the IP of the machine on which the funnel was active. This stopped working today. By changing this to something like:

```
"nodeAttrs": [
    {"target": ["tag:SOME-TAG"], "attr": ["funnel"]}
],
```

it seems to be working now.

**Comment by MightyDjinn** [comment]
Agree with the above comment, thanks for the fix [nkbookth](https://github.com/tailscale/tailscale/issues/11849#issuecomment-2211623437). This needs to further explained in the docs and possibly put directly into the blog entry. I had previously set the nodeAttrs for k8s and probably would have never found this.

**Comment by H0mire** [comment]
@nkbookth Thanks — toggling the funnel nodeAttr fixed it for me too, even though my ACL was already using "target": ["autogroup:member"]. :D

So the fix isn't specifically "IP → tag" — it's that making the funnel cap go away and come back re-triggers the control plane to recreate the public DNS record. Symptom: "Funnel on", cap present, works inside the tailnet, but public DNS = NXDOMAIN.

What worked: set "nodeAttrs": [], Save, then restore the original block, Save.  :D

**Comment by GabinSMD** [comment]
Same symptom here, and I can add a measurement that refines it: **the public DNS record is not simply missing — it is served inconsistently across the authoritative name servers.**

Toggling the `funnel` nodeAttr (thanks @nkbooth / @H0mire) is indeed the workaround. But before doing that, here is what the DNS actually looked like, which I think explains why several people in this thread describe the failure as intermittent rather than absolute.

### Measurement

`ts.net` has three authoritative servers. For the same funnel hostname, at the same moment, they disagree — and `ns1` even disagrees with itself depending on the transport:

| Queried | Result |
|---|---|
| `ns1.dnsimple.com` over **UDP** | `NXDOMAIN` (5/5 attempts) |
| `ns1.dnsimple.com` over **TCP** | `NOERROR`, 3 A records |
| `ns2.dnsimple-edge.net` over UDP | `NOERROR`, 3 A records |
| `ns3.dnsimple.com` over UDP | `NOERROR`, 3 A records |

```
dig     @ns1.dnsimple.com      penny.<tailnet>.ts.net A   # NXDOMAIN
dig +tcp @ns1.dnsimple.com     penny.<tailnet>.ts.net A   # NOERROR, 3 answers
dig     @ns2.dnsimple-edge.net penny.<tailnet>.ts.net A   # NOERROR, 3 answers
dig     @ns3.dnsimple.com      penny.<tailnet>.ts.net A   # NOERROR, 3 answers
```

The records that *do* come back point at the Funnel ingress, so the control plane clearly created them:

```
penny.<tailnet>.ts.net. 300 IN A    176.58.90.46 / .63 / .145
penny.<tailnet>.ts.net. 300 IN AAAA 2a00:dd80:3e::47e / ::b3a / ::e2c
```

A useful cross-check while diagnosing: on `ns1` over UDP, an `A` query returns `NXDOMAIN` while an `ANY` query for the very same name returns all 6 records. That inconsistency is what put me on the right track — `dig +short` had shown nothing, which reads identically to "no such record".

### Why this makes the failure look random

Because one authoritative server answers `NXDOMAIN`, any recursive resolver that happens to query it caches a **negative** answer for the zone's negative TTL. The result depends entirely on which resolver the client uses. Measured 30 minutes apart on the same hostname:

| Public resolver | Result |
|---|---|
| `8.8.8.8`, `8.8.4.4`, `9.9.9.9` | `NOERROR`, 3 answers |
| `1.1.1.1`, `1.0.0.1` | `NXDOMAIN` |

So for the affected users it is not "funnel is off", it is "funnel is reachable from some networks and not others, and it changes over time" — which matches several reports above about it working and then stopping.

### Node side was correct throughout

- `tailscale serve status --json` → `AllowFunnel: {"<host>:443": true}`, handler proxying to the local port
- `tailscale status --json` → node capabilities include `funnel` and `funnel-ports?ports=443,8443,10000`; `CertDomains` contains the hostname; `Health` empty
- `tailscale lock status` → tailnet lock not enabled
- `crt.sh` → 4 certificates issued for the hostname, the most recent still valid, so the public path had worked before
- restarting `tailscaled` logged `Hostinfo.IngressEnabled changed to true` but changed nothing in DNS, as expected once the cause is known

Version: `1.102.3` on Debian arm64 (Raspberry Pi 4).

Happy to share the exact tailnet hostname privately if a maintainer wants to inspect the records for this specific tailnet — I have left it out of the public thread.

**Comment by ad2012** [comment]
@GabinSMD I can confirm exactly the same behavior on _multiple_ accounts. There is also an open support ticket for this issue. This looks very much like a publication issue with Tailscale's public DNS.
